pkg/pipeline/input/sdk/synchronizer.go

```diff
@@ -1,39 +1,197 @@
+
+	firstPacket   int64
+	ntpStart      time.Time
+	endedAt       int64
+	syncInfo      map[uint32]*trackInfo
+	senderReports map[uint32]*rtcp.SenderReport
 }
 
-func (c *synchronizer) GetOrSetStartTime(t int64) int64 {
-	if c.startTime.CompareAndSwap(0, t) {
-		return t
+type trackInfo struct {
+	sync.Mutex
+
+	trackID    string
+	firstRtpTS int64
+	ptsDelay   int64
+	clockRate  int64
+	maxRTP     int64
+}
+
+func newSynchronizer() *synchronizer {
+	return &synchronizer{
+		syncInfo:      make(map[uint32]*trackInfo),
+		senderReports: make(map[uint32]*rtcp.SenderReport),
 	}
+}
 
-	startTime := c.startTime.Load()
-	c.delay.CompareAndSwap(0, t-startTime)
-	return startTime
+// initialize track sync info
+func (c *synchronizer) addTrack(track *webrtc.TrackRemote) {
+	c.Lock()
+	c.syncInfo[uint32(track.SSRC())] = &trackInfo{
+		trackID:   track.ID(),
+		clockRate: int64(track.Codec().ClockRate),
+	}
+	c.Unlock()
 }
 
-func (c *synchronizer) GetStartTime() int64 {
-	return c.startTime.Load()
+// initialize offsets once the first rtp packet is received
+func (c *synchronizer) firstPacketForTrack(pkt *rtp.Packet) {
+	now := time.Now().UnixNano()
+
+	c.Lock()
+	if c.firstPacket == 0 {
+		c.firstPacket = now
+	}
+	t := c.syncInfo[pkt.SSRC]
+	c.Unlock()
+
+	t.Lock()
+	t.firstRtpTS = int64(pkt.Timestamp)
+	t.ptsDelay = now - c.firstPacket
+	t.Unlock()
 }
 
-func (c *synchronizer) SetEndTime(t int64) {
-	c.endTime.Store(t)
+func (c *synchronizer) getStartedAt() int64 {
+	c.RLock()
+	defer c.RUnlock()
+
+	return c.firstPacket
 }
 
-func (c *synchronizer) GetEndTime() int64 {
-	return c.endTime.Load()
+func (c *synchronizer) onRTCP(packet rtcp.Packet) {
+	switch pkt := packet.(type) {
+	case *rtcp.SenderReport:
+		c.Lock()
+		defer c.Unlock()
+
+		if c.endedAt != 0 {
+			return
+		}
+
+		c.senderReports[pkt.SSRC] = pkt
+		if c.ntpStart.IsZero() {
+			if len(c.senderReports) < len(c.syncInfo) {
+				// wait for at least one report per track
+				return
+			}
+
+			// get the max ntp start time for all tracks
+			var minNTPStart time.Time
+			ntpStarts := make(map[uint32]time.Time)
+			for _, report := range c.senderReports {
+				t := c.syncInfo[report.SSRC]
+				pts, _ := t.getPTS(int64(report.RTPTime))
+				ntpStart := mediatransportutil.NtpTime(report.NTPTime).Time().Add(-pts)
+				if minNTPStart.IsZero() || ntpStart.Before(minNTPStart) {
+					minNTPStart = ntpStart
+				}
+				ntpStarts[report.SSRC] = ntpStart
+			}
+			c.ntpStart = minNTPStart
+
+			// update pts delay so all ntp start times match
+			for ssrc, ntpStart := range ntpStarts {
+				t := c.syncInfo[ssrc]
+				if diff := ntpStart.Sub(minNTPStart); diff != 0 {
+					t.Lock()
+					t.ptsDelay += int64(diff)
+					t.Unlock()
+				}
+			}
+		} else {
+			t := c.syncInfo[pkt.SSRC]
+			pts, _ := t.getPTS(int64(pkt.RTPTime))
+			expected := mediatransportutil.NtpTime(pkt.NTPTime).Time().Sub(c.ntpStart)
+			if pts != expected {
+				diff := expected - pts
+				if diff > time.Millisecond*20 {
```

Comment:
Can it happen that pts > expected? If so, should we consider abs(diff)?

Reply:
ah, yes. Happens frequently